crates/ty_python_semantic/resources/mdtest/generics/pep695/variance.md

```diff
@@ -446,4 +398,101 @@ static_assert(not is_gradual_equivalent_to(D[Any], C[A]))
+class B(A): ...
+
+class C[X]:
+    def f(self) -> "D[X]":
+        return D()
+
+    def g(self, x: X) -> None: ...
+
+class D[Y]:
+    def h(self) -> C[Y]:
+        return C()
+```
+
+`X` is contravariant in both `C` and `D`:
+
+- `C` has two occurrences of `X`
+    - `X` occurs in the return type of `f` as `D[X]`
+        - `D` has one occurrence of `X`
+            - `X` occurs in the return type of `h` as `C[X]`
+    - `X` occurs contravariantly as a parameter in `g`
+
+We want the _greatest_ inferrable variance, so we start by assuming `X` occurs bivariantly in `C`.
+Therefore, `X` has one bivariant occurrence, and one contravariant occurrence, which resolves to
+contravariant.
+
+Then because `X` is contravariant in `C` `D` is as well.
+
+```py
+static_assert(not is_subtype_of(C[B], C[A]))
+static_assert(is_subtype_of(C[A], C[B]))
+static_assert(not is_subtype_of(C[A], C[Any]))
+static_assert(not is_subtype_of(C[B], C[Any]))
+static_assert(not is_subtype_of(C[Any], C[A]))
+static_assert(not is_subtype_of(C[Any], C[B]))
+
+static_assert(not is_subtype_of(D[B], D[A]))
+static_assert(is_subtype_of(D[A], D[B]))
+static_assert(not is_subtype_of(D[A], D[Any]))
+static_assert(not is_subtype_of(D[B], D[Any]))
+static_assert(not is_subtype_of(D[Any], D[A]))
+static_assert(not is_subtype_of(D[Any], D[B]))
+```
+
+## Type Variable Closure
+
+```py
+from ty_extensions import is_assignable_to, is_equivalent_to, is_gradual_equivalent_to, is_subtype_of, static_assert, Unknown
+from typing import Any
+
+class A: ...
+class B(A): ...
+
+class C[T]:
+    class D:
+        def f(self, value: T) -> None: ...
+
+    def f(self, value: D) -> None: ...
+```
+
+When inferring variance, we can't skip over classes just because they don't have type variables.
+This is because they can close over type variables, such as `D` in the above example.
+
+In this case, `D` is contravariant in `T`. `C` has the contravariant occurence _within_ `D` and a
+contravariant occurrence _of_ `D`. The latter occurrence is ultimately thus covariant (contravariant
+occurrence in contravariant position). Then `C` has both a covariant and a contravariant occurrence
+of `D`, so it is invariant.
+
+TODO: the bottom set of these are failing, perhaps something to do with subclasses of specialized
+aliases
+
+```py
+static_assert(not is_subtype_of(C[B], C[A]))
+static_assert(not is_subtype_of(C[A], C[B]))
+static_assert(not is_subtype_of(C[A], C[Any]))
+static_assert(not is_subtype_of(C[B], C[Any]))
+static_assert(not is_subtype_of(C[Any], C[A]))
+static_assert(not is_subtype_of(C[Any], C[B]))
+
+static_assert(not is_subtype_of(C[B].D, C[A].D))
+static_assert(is_subtype_of(C[A].D, C[B].D))
+static_assert(not is_subtype_of(C[A].D, C[Any].D))
+static_assert(not is_subtype_of(C[B].D, C[Any].D))
+static_assert(not is_subtype_of(C[Any].D, C[A].D))
+static_assert(not is_subtype_of(C[Any].D, C[B].D))
+```
+
+[linear-time-variance-talk]: https://www.youtube.com/watch?v=7uixlNTOY4s&t=9705s
```

Comment:
We should see if Martin has a linkable copy of his slides that we can cite as well.